src/util/shmem/fd_shmem_user.c

```diff
@@ -74,6 +75,46 @@ fd_shmem_private_map_query_by_addr( fd_shmem_join_info_t * map,
   return def;
 }
 
+static int
+fd_shmem_private_grab_region( ulong addr,
+                              ulong size ) {
+  void *mmap_ret;
+
+  mmap_ret = mmap( (void*)addr, size, PROT_READ, MAP_ANON|MAP_PRIVATE, -1, 0 );
+
+  if( mmap_ret == MAP_FAILED )
+    return 0;
+
+  if( (ulong)mmap_ret != addr ) {
+    munmap( mmap_ret, size );
+    return 0;
+  }
+
+  return 1;
+}
+
+static ulong
+fd_shmem_private_get_random_mappable_addr( ulong size ) {
```

Comment:
From the man page for `MAP_FIXED`:

```
If  the memory region specified by addr and len overlaps pages of any existing mapping(s), then the overlapped part of the existing mapping(s) will be discarded. [...] Furthermore, this option is extremely hazardous (when used on its own), because it forcibly removes preexisting mappings, making it easy for a multithreaded process to corrupt its own address space.
```

Is this right? Do you need to keep rolling if you conflict with an existing mapping? (I guess there's basically no chance with a true RNG for 64 bits).

I took a look at the chrome randomizer and it gave a few pointers as well,

https://github.com/chromium/chromium/blob/abe2b94097458ae7e3741706b310aa70c4e7ee07/base/allocator/partition_allocator/src/partition_alloc/address_space_randomization.h#L146

We want to disable for ASAN builds. And I'm not sure if the address space we use matches

```
      // Linux (and macOS) support the full 47-bit user space of x64 processors.
      // Use only 46 to allow the kernel a chance to fulfill the request.
      PA_ALWAYS_INLINE constexpr uintptr_t ASLRMask() {
        return AslrMask(46);
      }
      PA_ALWAYS_INLINE constexpr uintptr_t ASLROffset() {
        return AslrAddress(0);
      }
```

They also just mask off the bits outside bounds rather than repeatedly rolling.

```
  uintptr_t random = static_cast<uintptr_t>(internal::RandomValue());

#if BUILDFLAG(HAS_64_BIT_POINTERS)
  random <<= 32ULL;
  random |= static_cast<uintptr_t>(internal::RandomValue());

  // The ASLRMask() and ASLROffset() constants will be suitable for the
  // OS and build configuration.
  random &= internal::ASLRMask();
  random += internal::ASLROffset();
```